python/custom_test_files.py

```diff
@@ -0,0 +1,189 @@
+        df.reset_index()
+        row_count = len(df.index)
+        #Creates a new list of LEI's for the ULI column by bank.
+        if bank == 'Bank0': 
+            unique_lei_for_uli = {'uli':list(['B90YWS6AFX2LGWOXJ1LD']*int((row_count)))}
+        
+        if bank == 'Bank1':
+            unique_lei_for_uli = {'uli':list(['BANK1LEIFORTEST12345']*int((row_count)))}
+        
+        uli_col = pd.DataFrame(unique_lei_for_uli) #Creates a dataframe of new ULI's.
+        #Amends the ULI's by adding up to 23 additional characters. 
+        uli_col['uli'] = uli_col.apply(lambda x: x.uli + lg.char_string_gen(length = 23), axis=1)
+        #Adds a checkdigit to the end of the ULI. 
+        uli_col['uli'] = uli_col.apply(lambda x: x.uli + lg.check_digit_gen(ULI=x.uli), axis=1)
+        #Checks whether ULI's are unique using the uli_check function. 
+        
+        if self.uli_check(col = uli_col) is True:
+            df['uli'] = uli_col['uli'] #If unique, ULI's in the dataframe are replaced. 
+            print("Unique ULIs Assigned for " + str(bank))
+            return df
+        else: 
+            #If not unique, the assign_uli function is re-run.
+            print("Re-Running")
+            assign_uli(self, ts_file=ts_file, df=df)  
+    
+    def uli_check(self, col=None):
+        """
+        Passes in a single column dataframe to determine whether values within the column
+        are unique. 
+        
+        Returns a True statement if they are, and returns a False statement if not. 
+        """
+        col['unique'] = col
+        if len(col['unique'].unique()) == len(col.index): #Checks whether column values are unique. 
+            return True
+        else: 
+            return False
+        
+    def lar_to_hmda_file(self, df=None, ts_file=None, save_file=None):
+        
+        """
+        Creates a HMDA submission file, passing in a dataframe of LAR rows, the TS file, and a filepath to store
+        the file. 
+        
+        """
+        
+        #Store the number of rows in the lar_rows dataframe. 
+        row_count = len(df.index)
+        
+        #Replaces the lar_data text file with the lar_rows dataframe.
+        df.to_csv("../edits_files/file_parts/lar_data.txt", sep="|", index=False, header=None)
+
+
+        #Reads in TS file.
+        ts_df = pd.read_csv(ts_file, delimiter="|", header=None, dtype='object',
+                         keep_default_na=False)
+        
+        #Stores the second column of the TS file as 'bank'. 
+        bank = ts_df.iloc[0][1]
+        
+        #Creates a new file path to place the new HMDA submission file. 
+        os.makedirs(os.path.dirname(save_file), exist_ok=True)
+        
+        #Copies the TS text file into the save filepath specified in the function and renames it. 
+        sh.copyfile(ts_file, save_file)
+        
+        #Reads in lar_data text file and appends it to the Transmittal Sheet row in the new file. 
```

Comment:
You should be able to store the LAR data object you've created through concatenation in memory using a self.lar_data type argument. This makes the variable visible inside the class and eliminates the need to save the LAR data as a file which will increase your performance and reduce file clutter.